File: src/core/cfr_plus.py
import numpy as np
from numba import jit
import json
from typing import Dict, List

class CFRPlus:
    def __init__(self, actions: int):
        self.regret = np.zeros(actions, dtype=np.float32)
        self.strategy_sum = np.zeros(actions, dtype=np.float32)
        self.actions = actions
        
    @jit(nopython=True, nogil=True)
    def _update_regret(self, regret: np.ndarray, utility: np.ndarray, strategy: np.ndarray) -> None:
        regret += utility - np.dot(strategy, utility)
        
    def update(self, iteration: int, utilities: np.ndarray) -> np.ndarray:
        strategy = self.get_strategy()
        self._update_regret(self.regret, utilities, strategy)
        self.strategy_sum += strategy * iteration
        return strategy
        
    @jit(nopython=True)
    def get_strategy(self) -> np.ndarray:
        pos_regret = np.maximum(self.regret, 0)
        normalizer = np.sum(pos_regret)
        return pos_regret / normalizer if normalizer > 0 else np.ones(self.actions) / self.actions
        
    def save_strategy(self, filepath: str) -> None:
        np.savez_compressed(filepath, strategy=self.strategy_sum)
        
    @classmethod
    def load_strategy(cls, filepath: str) -> 'CFRPlus':
        data = np.load(filepath)
        instance = cls(data['strategy'].shape[0])
        instance.strategy_sum = data['strategy']
        return instance

class GameTree:
    def __init__(self, config_path: str):
        with open(config_path) as f:
            self.config = json.load(f)
        self.nodes = {}
        
    def get_node(self, infoset: str) -> CFRPlus:
        if infoset not in self.nodes:
            self.nodes[infoset] = CFRPlus(len(self.config['actions']))
        return self.nodes[infoset]

File: src/core/cfr_plus.js
import { Random } from '../utils/random.js';

class CFRPlus {
  constructor(actions) {
    this.regret = new Array(actions).fill(0);
    this.strategySum = new Array(actions).fill(0);
    this.actions = actions;
    this.random = new Random('poker-ai'); // Seeded random number generator
  }

  update(iteration, utilities) {
    const strategy = this.getStrategy();
    for (let i = 0; i < this.actions; i++) {
      this.regret[i] += utilities[i] - strategy.reduce((sum, s, j) => sum + s * utilities[j], 0);
    }
    for (let i = 0; i < this.actions; i++) {
      this.strategySum[i] += strategy[i] * iteration;
    }
    return strategy;
  }

  getStrategy() {
    const posRegret = this.regret.map(r => Math.max(r, 0));
    const normalizer = posRegret.reduce((sum, r) => sum + r, 0);
    return normalizer > 0
      ? posRegret.map(r => r / normalizer)
      : new Array(this.actions).fill(1 / this.actions);
  }

    saveStrategy(filepath) {
        const fs = require('fs');
        fs.writeFileSync(filepath, JSON.stringify(this.strategySum));
    }
    
    static loadStrategy(filepath) {
        const fs = require('fs');
        try {
            const data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
            const instance = new CFRPlus(data.length);
            instance.strategySum = data;
            return instance;
        } catch (e) {
            console.warn("Failed to load strategy, returning a blank one", e);
            return null;
        }
    }
}


class GameTree {
    constructor(config) {
        this.config = config;
        this.nodes = {};
    }

    get_node(infoset) {
        if (!(infoset in this.nodes)) {
            this.nodes[infoset] = new CFRPlus(this.config.actions.length);
        }
        return this.nodes[infoset];
    }
}

export { CFRPlus, GameTree };
